src/types/complex.rs

```diff
@@ -17,205 +19,191 @@ pyobject_native_type!(
 );
 
 impl PyComplex {
-    /// Creates a new Python `complex` object, from its real and imaginary values.
+    /// Creates a new `PyComplex` from the given real and imaginary values.
     pub fn from_doubles(py: Python, real: c_double, imag: c_double) -> &PyComplex {
         unsafe {
             let ptr = ffi::PyComplex_FromDoubles(real, imag);
             py.from_owned_ptr(ptr)
         }
     }
     /// Returns the real part of the complex number.
+    #[must_use = "method returns a new number and does not mutate the original value"]
```

Comment:
Is this `#[must_use]` really necessary? We could make a similar argument for all "getter" methods, but I don't see `#[must_use]` on most of them...

Reply:
Right, real/imag don't actually return a PyComplex. I'll remove them. How do you feel about must_use on `.pow()`?

Reply:
I've tended to treat `#[must_use]` as "it's a bad bug if the result of this is not used". The existing two uses in PyO3, in `GilGuard` and `IntoPyPointer::into_ptr` are in the spirit of this.

In https://github.com/rust-lang/rust/issues/48926 the stdlib itself is trying to decide on the appropriate use of this attribute. I think I'd rather we stayed away from overuse of it until there's been consensus in that issue.

(I note that `f64::pow` is `#[must_use]`, but `num_complex::Complex::powf` is not.)

Did you come across a real-world bug arising from discarding the return value of these functions?

Reply:
Imho, must_use is appropriate if not using the return value makes the entire operation useless.

For example, I would argue that something like this is incorrect - it contains dead code and/or won't do what the author intends. Therefore a warning is appropriate.

```rust
fn main() {
    let mut i = 42_usize;
    i += 1;
    i.pow(2);
    println!("{:?}", i)
}
```
I'm happy to remove them from the methods that don't return `Self`. I'll remove them from `pow` if you feel strongly about it (?)

> Did you come across a real-world bug arising from discarding the return value of these functions?

Not these in particular; but I would have been burned by not handling `Result<()>` a few times, if it wasn't for that having a lint.

Reply:
I think for getters it's overkill to apply this, especially when it is obvious that the method is a getter.

Personally, I tend to apply must_use only in cases where not using the return value can cause a hard-to-find bug. I.e. with `pow()`, if you test your code at least once, you'll find the bug.  For something returning `Result`, you won't if you only test the happy path.

Reply:
@birkenfeld's description of `#[must_use]` is pretty much exactly the same as my own view of it.

Overall I'm slightly leaning towards not adding it. I'd be happy to reverse that position if appropriate once `std` has settled on a convention how to use the attribute.

If you feel strongly that it should be there, I can be persuaded. My reservation if that if it doesn't fall into the convention `std` eventually sets then we'll want to remove it later.